azure-cli-diff-tool/azure_cli_diff_tool/version_upgrade.py

```diff
@@ -0,0 +1,174 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# -----------------------------------------------------------------------------
+# pylint: disable=line-too-long
+
+# NOTE: The version uodate rules in this doc is complied with the doc below
+# https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md
+
+from packaging.version import parse
+from ._const import (PREVIEW_INIT_SUFFIX, VERSION_MAJOR_TAG, VERSION_MINOR_TAG,
+                     VERSION_PATCH_TAG, VERSION_STABLE_TAG, VERSION_PREVIEW_TAG, VERSION_PRE_TAG,
+                     CLI_EXTENSION_INDEX_URL)
+from .utils import DiffLevel
+
+
+class ModuleVersion:
+
+    def __init__(self, parse_version):
+        self.major = parse_version.major
+        self.minor = parse_version.minor
+        self.patch = parse_version.micro
+        self.pre = parse_version.pre
+        self.pre_num = parse_version.pre and parse_version.pre[1]
+
+    def __str__(self):
+        version_arr = [self.major, ".", self.minor, ".", self.patch]
+        if self.pre:
+            version_arr.append(self.pre)
+            version_arr.append(self.pre_num)
+        return "".join([str(item) for item in version_arr])
+
+
+class VersionUpgradeMod:
+
+    def __init__(self, module_name, current_version, is_preview, is_experimental,
```

Comment:
Is there any reason to keep `is_experimental`?

Reply:
This is passed from current raw metadata and used to check whether to insert isPreview tag to next extension version for old cli users.